gstd/src/msg.rs

```diff
@@ -17,155 +17,87 @@
-        pub fn gr_reply_push(data_ptr: *const u8, data_len: u32);
-        pub fn gr_reply_to(dest: *mut u8);
-        pub fn gr_send(
-            program: *const u8,
-            data_ptr: *const u8,
-            data_len: u32,
-            gas_limit: u64,
-            value_ptr: *const u8,
-            message_id_ptr: *mut u8,
-        );
-        pub fn gr_send_commit(
-            handle: u32,
-            message_id_ptr: *mut u8,
-            program: *const u8,
-
-            gas_limit: u64,
-            value_ptr: *const u8,
-        );
-        pub fn gr_send_init() -> u32;
-        pub fn gr_send_push(handle: u32, data_ptr: *const u8, data_len: u32);
-        pub fn gr_size() -> u32;
-        pub fn gr_source(program: *mut u8);
-        pub fn gr_value(val: *mut u8);
-        pub fn gr_wait() -> !;
-        pub fn gr_wake(waker_id_ptr: *const u8);
+pub use gcore::msg::{gas_available, id, reply_to, source, value, wait, wake};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MessageHandle(gcore::MessageHandle);
+
+impl MessageHandle {
+    pub fn init() -> Self {
+        send_init()
+    }
+
+    pub fn push<E: Encode>(&self, payload: E) {
+        gcore::msg::send_push(&self.0, &payload.encode())
+    }
+
+    pub fn commit(self, program: ProgramId, gas_limit: u64, value: u128) -> MessageId {
+        gcore::msg::send_commit(self.0, program, gas_limit, value)
     }
 }
 
-pub fn gas_available() -> u64 {
-    unsafe { sys::gr_gas_available() }
+impl Output for MessageHandle {
+    fn write(&mut self, bytes: &[u8]) {
+        gcore::msg::send_push(&self.0, bytes)
+    }
 }
 
-pub fn id() -> MessageId {
-    let mut msg_id = MessageId::default();
-    unsafe { sys::gr_msg_id(msg_id.0.as_mut_ptr()) }
-    msg_id
+pub fn load<D: Decode>() -> Result<D, codec::Error> {
+    D::decode(&mut gcore::msg::load().as_ref())
 }
 
-pub fn load() -> Vec<u8> {
-    unsafe {
-        let message_size = sys::gr_size() as usize;
-        let mut data = Vec::with_capacity(message_size);
-        data.set_len(message_size);
-        sys::gr_read(0, message_size as _, data.as_mut_ptr() as _);
-        data
-    }
+pub fn load_bytes() -> Vec<u8> {
+    gcore::msg::load()
 }
 
-pub fn reply(payload: &[u8], gas_limit: u64, value: u128) {
-    unsafe {
-        sys::gr_reply(
-            payload.as_ptr(),
-            payload.len() as _,
-            gas_limit,
-            value.to_le_bytes().as_ptr(),
-        )
-    }
+pub fn reply<E: Encode>(payload: E, gas_limit: u64, value: u128) {
+    let bytes = payload.encode();
+    gcore::msg::reply(&bytes, gas_limit, value)
+}
+
+pub fn reply_bytes(payload: &[u8], gas_limit: u64, value: u128) {
+    gcore::msg::reply(payload, gas_limit, value)
 }
 
-pub fn reply_push(payload: &[u8]) {
-    unsafe { sys::gr_reply_push(payload.as_ptr(), payload.len() as _) }
+pub fn reply_push<E: Encode>(payload: E) {
+    let bytes = payload.encode();
+    gcore::msg::reply_push(&bytes)
```

Comment:
No need to return the unit type result.
```suggestion
    gcore::msg::reply_push(&bytes);
```